component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/voiceotp/VoiceOTPAuthenticator.java

```diff
@@ -0,0 +1,2034 @@
+        }
+    }
+
+    /**
+     * Get MultiOptionURI query parameter from the request.
+     *
+     * @param request Http servlet request.
+     * @return MultiOptionURI query parameter.
+     */
+    private String getMultiOptionURIQueryParam(HttpServletRequest request) {
+
+        if (request != null && StringUtils.isNotEmpty(request.getParameter(VoiceOTPConstants.MULTI_OPTION_URI))) {
+            return "&" + VoiceOTPConstants.MULTI_OPTION_URI + "="
+                        + Encode.forUriComponent(request.getParameter(VoiceOTPConstants.MULTI_OPTION_URI));
+        }
+        return StringUtils.EMPTY;
+    }
+
+    /**
+     * Get the mobile number from user's profile to send an otp.
+     *
+     * @param request     The HttpServletRequest.
+     * @param response    The HttpServletResponse.
+     * @param context     The AuthenticationContext.
+     * @param username    The Username.
+     * @param queryParams The queryParams.
+     * @return the mobile number
+     * @throws AuthenticationFailedException
+     * @throws VoiceOTPException
+     */
+    private String getMobileNumber(HttpServletRequest request, HttpServletResponse response,
+                                   AuthenticationContext context, String username,
+                                   String queryParams) throws AuthenticationFailedException, VoiceOTPException {
+
+        String mobileNumber = VoiceOTPUtils.getMobileNumberForUsername(username);
+        if (StringUtils.isEmpty(mobileNumber)) {
+            String requestMobile = request.getParameter(VoiceOTPConstants.MOBILE_NUMBER);
+            if (StringUtils.isBlank(requestMobile) && !isMobileNumberUpdateFailed(context) && isCodeMismatch(context)) {
+                mobileNumber = String.valueOf(context.getProperty(VoiceOTPConstants.REQUESTED_USER_MOBILE));
+            } else if (StringUtils.isBlank(requestMobile)) {
+                if (log.isDebugEnabled()) {
+                    log.debug("User has not registered a mobile number: " + username);
+                }
+                redirectToMobileNoReqPage(response, context, queryParams);
+            } else {
+                context.setProperty(VoiceOTPConstants.REQUESTED_USER_MOBILE, requestMobile);
+                mobileNumber = requestMobile;
+            }
+        }
+        return mobileNumber;
+    }
+
+    /**
+     * Get the loginPage from authentication.xml file or use the login page from constant file.
+     *
+     * @param context The AuthenticationContext
+     * @return The loginPage
+     */
+    private String getLoginPage(AuthenticationContext context) {
+
+        String loginPage = VoiceOTPUtils.getLoginPageFromXMLFile(context);
+        if (StringUtils.isEmpty(loginPage)) {
+            loginPage = ConfigurationFacade.getInstance().getAuthenticationEndpointURL()
+                    .replace(VoiceOTPConstants.LOGIN_PAGE, VoiceOTPConstants.VOICE_LOGIN_PAGE);
+
+            log.debug("Default authentication endpoint context is used");
```

Comment:
Shall we add more context to this debug log if we really need this? Having this message only is not meaningful